### PR TITLE
[MPS] Pixel shuffle unshuffle support

### LINKED ISSUE #83196: PixelShuffle is very slow on MPS compared to ConvTranspose2d and PixelShuffle on cuda
### 🐛 Describe the bug

I changed the decode layers of a CNN to use pixel shuffle instead of transpose_conv2d. The model immediately took 3x longer to run. GPU usage was still high and there was no warning about fallback to using the CPU. I recently starting training the CNN remotely using coda and saw a reduction in training by 66%. The only conclusion is that there is a problem with the MPS implementation.

The commented out lines show the ConvTranspose2D  being replaced with the Conv2d followed by PixelShuffle

class decodeUpBlock(nn.Module):
    def __init__(self, inFM, outFM, kSize, outSize, stride=1, padding=0, output_padding=0):
        super(decodeUpBlock, self).__init__()
        self.layers = nn.Sequential()
        #self.layers.add_module("TConv", nn.ConvTranspose2d(inFM,outFM,kSize,stride=stride,padding=padding,output_padding=output_padding))
        self.layers.add_module("Conv2", nn.Conv2d(inFM,4*outFM,kSize,stride=1,padding=padding))
        self.layers.add_module("pShfl", nn.PixelShuffle(stride))
        self.layers.add_module("lNorm", nn.LayerNorm([outFM,outSize,outSize]))
        self.layers.add_module("ReLU", nn.LeakyReLU(0.2,inplace=True))

### Versions

Collecting environment information...
PyTorch version: 1.12.0.post2
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.0 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.5 | packaged by conda-forge | (main, Jun 14 2022, 07:07:06) [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-13.0-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] torch==1.12.0.post2
[pip3] torchvision==0.13.0a0
[conda] numpy                     1.23.1          py310h0a343b5_0    conda-forge
[conda] pytorch                   1.12.0          cpu_py310h911b1ea_2    conda-forge
[conda] torchvision               0.13.0          cpu_py310he68663e_0    conda-forge

### PR DESCRIPTION
**Pull request by alexdremov**
Fixes #83196

Now, MPS implementation is blazingly fast.

Though, I have several questions on improving this PR:

1. I copied code from `test_nn.py`. Is there better way to test this?
2. I decided to use `usepixelshuffleorder:YES`. Am I right performance-wise? According to docs:
```
`usePixelShuffleOrder` can be
used to control how the data within spatial blocks is ordered in the
`depthAxis` dimension: with `usePixelShuffleOrder=YES` the values within the
spatial blocks are stored contiguosly within the `depthAxis` dimension whereas
otherwise they are stored interleaved with existing values in the `depthAxis` dimension.
```

cc: @razarmehr @kulinseth

### CODE REVIEW

File: aten/src/ATen/native/PixelShuffle.h
Comment (alexdremov) [inline aten/src/ATen/native/PixelShuffle.h:1]: Moved it to separate header as theese checks are needed in both CPU and MPS code

**Comment by alexdremov** [comment]
This fails on MacOS 12. Any ideas? Should I just make MPS implementation unavailable on MacOS 12?

**Review by kulinseth — COMMENTED** [review]
Looks great . Can you please refactor the upscale and downscale code for pixel shuffle and unshuffle.

**Comment by kulinseth** [comment]
> This fails on MacOS 12. Any ideas? Should I just make MPS implementation unavailable on MacOS 12?

You can use fallback for macOS 12 . The api should work but seems like there are bugs ..

File: aten/src/ATen/native/mps/operations/PixelShuffle.mm [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+using namespace at::mps;
+
+namespace at::native {
+
+Tensor pixel_shuffle_mps(const Tensor& self, int64_t upscale_factor) {
+  using namespace mps;
+  using CachedGraph = MPSUnaryCachedGraph;
+
+  if (upscale_factor == 1) {
+    return self;
```
Comment (alexdremov) [inline aten/src/ATen/native/mps/operations/PixelShuffle.mm:13]: @kulinseth should this return `self.clone()`? What is expected behaviour?

Comment (alexdremov) [reply aten/src/ATen/native/mps/operations/PixelShuffle.mm:22]: Figured out it should cause we don.t want this:

```python
t = torch.randn(1, 4, 3, 3, device='mps', requires_grad=True)
res = F.pixel_shuffle(t, 1)
t -= 10000  # res modified too
```

**Comment by alexdremov** [comment]
@kulinseth any updates?

File: aten/src/ATen/native/mps/operations/PixelShuffle.mm [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  const int64_t upscale_factor_squared = upscale_factor * upscale_factor;
+  const int64_t oc = c / upscale_factor_squared;
+  const int64_t oh = h * upscale_factor;
+  const int64_t ow = w * upscale_factor;
+
+  std::vector<int64_t> out_shape(self.sizes().begin(), self_sizes_batch_end);
+  out_shape.insert(out_shape.end(), {oc, oh, ow});
+
+  Tensor output = at::empty(out_shape, self.options());
+
```
Comment (kulinseth) [inline aten/src/ATen/native/mps/operations/PixelShuffle.mm:45]: I think this can be added...

```  
  auto output = at::empty({0}, self.options());
  if (output.numel() == 0) {
    return output;
  }
```

**Review by kulinseth — APPROVED** [review]
Modulo this looks good.

**Comment by kulinseth** [comment]
> > Looks great . Can you please refactor the upscale and downscale code for pixel shuffle and unshuffle.
> 
> Could you please clarify which code specifically? Is there something wrong in `PixelShuffle.mm`?

I meant if you start the upscale and downscale code, the calculations are same. The Graph and other parts can be refactored:

```
const int64_t c = self.size(-3);
   const int64_t h = self.size(-2);
   const int64_t w = self.size(-1);
   constexpr auto NUM_NON_BATCH_DIMS = 3;
   const auto self_sizes_batch_end = self.sizes().end() - NUM_NON_BATCH_DIMS;

   const int64_t downscale_factor_squared = downscale_factor * downscale_factor;
   const int64_t oc = c * downscale_factor_squared;
   const int64_t oh = h / downscale_factor;
   const int64_t ow = w / downscale_factor;

   std::vector<int64_t> out_shape(self.sizes().begin(), self_sizes_batch_end);
   out_shape.insert(out_shape.end(), {oc, oh, ow});
```

I would encourage you move all the code to a helper function and then call it from Shuffle and Unshuffle.

**Comment by alexdremov** [comment]
> ## Merge failed
> **Reason**: Approval needed from one of the following: YXIE14, dulinriley, Hangjun, eprivezentsev, govardhan, ...
> 
> Details for Dev Infra team
> Raised by [workflow job](https://github.com/pytorch/pytorch/actions/runs/4772105245)
> Failing merge rule: Core Maintainers

@kulinseth, could you tag an appropriate core maintainer, please? ;)

**Review by kulinseth — COMMENTED** [review]
@albanD can you please take a look ?

**Comment by kulinseth** [comment]
> > ## Merge failed
> > **Reason**: Approval needed from one of the following: YXIE14, dulinriley, Hangjun, eprivezentsev, govardhan, ...
> > Details for Dev Infra team
> > Raised by [workflow job](https://github.com/pytorch/pytorch/actions/runs/4772105245)
> > Failing merge rule: Core Maintainers
> 
> @kulinseth, could you tag an appropriate core maintainer, please? ;)

Done , added @albanD

**Comment by alexdremov** [comment]
@albanD, could you please take a look?

**Comment by alexdremov** [comment]
@albanD, @kulinseth this PR is almost a month old

**Comment by kulinseth** [comment]
> @albanD, @kulinseth this PR is almost a month old

Pinged @albanD.

File: aten/src/ATen/native/mps/operations/PixelShuffle.mm [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  return output;
+}
+
+Tensor pixel_shuffle_mps(const Tensor& self, int64_t upscale_factor) {
+  if (!is_macos_13_or_newer()) {
+    TORCH_WARN_ONCE("MPS: pixel_shuffle op is supported starting from macOS 13.0. ",
+                    "Falling back on CPU. This may have performance implications.");
+
+    return at::native::pixel_shuffle_cpu(self.to("cpu"), upscale_factor).clone().to("mps");
```
Comment (albanD) [inline aten/src/ATen/native/mps/operations/PixelShuffle.mm:105]: Why clone the output here and below?

File: test/test_mps.py [hunk 4]
```diff
@@ -934,6 +934,110 @@ def leak_gpu0():
         with self.assertRaisesRegex(RuntimeError, r"MPS driver API confirmed .+"):
             leak_gpu0()
 
+
+# These tests were taken from test/test_nn.py
```
Comment (albanD) [inline test/test_mps.py:938]: We have OpInfo based tests for this. So this should already properly covered?

Comment (albanD) [reply test/test_mps.py:938]: I meant there is a regular OpInfo: https://github.com/pytorch/pytorch/blob/f542b31c9dea1937328d3c3ce03738ce340faf1e/torch/testing/_internal/common_methods_invocations.py#L17985

That one should already be ran by the existing tests in this file no?

File: aten/src/ATen/native/mps/operations/PixelShuffle.mm (hunk 3 above)
Comment (alexdremov) [reply aten/src/ATen/native/mps/operations/PixelShuffle.mm:105]: True, unneeded

File: test/test_mps.py (hunk 4 above)
Comment (alexdremov) [reply test/test_mps.py:939]: If these tests already run, I'll be happy to remove them. Is there a way to check it?

Comment (albanD) [reply test/test_mps.py:939]: In the CI logs on the mps machine, you can look for `nn_functional_pixel_shuffle`

Comment (alexdremov) [reply test/test_mps.py:939]: > In the CI logs on the mps machine, you can look for `nn_functional_pixel_shuffle`

Found these lines

```
  test_mps.py::TestConsistencyCPU::test_output_grad_match_nn_functional_pixel_shuffle_cpu_float16 PASSED [0.1346s] [ 21%]
  test_mps.py::TestConsistencyCPU::test_output_grad_match_nn_functional_pixel_shuffle_cpu_float32 PASSED [0.1138s] [ 21%]
```

Comment (alexdremov) [reply test/test_mps.py:939]: But `test_pixel_shuffle_unshuffle` that I copied here is not presented elsewhere.

Indeed, `test_pixel_shuffle_unshuffle` in `test_nn.py` does not use `device` inside. So, I can't see how it can be automatically checked for MPS

Comment (albanD) [reply test/test_mps.py:939]: Do these already cover this test you just copied here? (and the test in test_nn is just redundant)
Or they test different things?

Comment (alexdremov) [reply test/test_mps.py:939]: They're different

Comment (malfet) [reply test/test_mps.py:939]: IMO it's a bid redundant, but can be done as a followup PR

File: aten/src/ATen/native/PixelShuffle.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
+#include <ATen/core/Tensor.h>
+#include <c10/util/Exception.h>
+
+namespace at {
+namespace native {
+
+inline void check_pixel_shuffle_shapes(const Tensor& self, int64_t upscale_factor) {
```
Comment (malfet) [inline aten/src/ATen/native/PixelShuffle.h:7]: This could have been done as structured kernel, but not specific to this PR.

**Review by malfet — APPROVED** [review]
Looks good to me, though can we please add check/guards for pixel shuffle using different dtypes (something tells me it's going to crash with integer tensors for example)
Also, what about non-contiguous tensors?

File: aten/src/ATen/native/mps/operations/PixelShuffle.mm [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    string key = (upscale ? "pixel_shuffle_" : "pixel_unshuffle_") + getTensorsStringKey({self}) + "_factor_" +
+        std::to_string(factor);
+    CachedGraph* cachedGraph = cache_->LookUpAs<CachedGraph>(key);
+    if (!cachedGraph) {
+      MPSCachedGraph* tmpCachedGraph = cache_->CreateCachedGraph(key, ^MPSCachedGraph*() {
+        CachedGraph* newCachedGraph = nil;
+
+        @autoreleasepool {
+          MPSGraph* mpsGraph = make_mps_graph();
+          newCachedGraph = new CachedGraph(mpsGraph);
```
Comment (malfet) [inline aten/src/ATen/native/mps/operations/PixelShuffle.mm:57]: Please consider using `LookUpOrCreateCachedGraph` template
```suggestion
    auto* cachedGraph = LookUpOrCreateCachedGraph<CachedGraph>([&](auto mpsGraph, auto newCachedGraph) {

```

File: test/test_mps.py (hunk 4 above)
Comment (kulinseth) [reply test/test_mps.py:939]: @alexdremov can you please remove the test_nn copied test in this PR to avoid duplication? We are planning to enable test_nn testing with 'mps' device soon and will have this enabled there.

**Comment by kulinseth** [comment]
@alexdremov , can we resurrect this PR ?

**Comment by alexdremov** [comment]
> @alexdremov , can we resurrect this PR ? 

@kulinseth Seems like code side is fine and it successfully rebased. 

There’s only @malfet questions that need to be checked. I

https://github.com/pytorch/pytorch/pull/99306#pullrequestreview-1429463238

**Comment by alexdremov** [comment]
@kulinseth I've ressurected this PR. Merging?

- I added non-contiguous tests
- This works with integers. Tests run checks for various dtypes:
```
test/test_mps.py::TestPixelShuffle::test_pixel_shuffle_unshuffle PASSED [5.7919s]                                                                                                                        [  4%]
test/test_mps.py::TestConsistencyCPU::test_output_grad_match_nn_functional_pixel_shuffle_cpu_float16 PASSED [0.0728s]                                                                                    [  9%]
test/test_mps.py::TestConsistencyCPU::test_output_grad_match_nn_functional_pixel_shuffle_cpu_float32 PASSED [0.0689s]                                                                                    [ 14%]
test/test_mps.py::TestConsistencyCPU::test_output_grad_match_nn_functional_pixel_unshuffle_cpu_float16 PASSED [0.0044s]                                                                                  [ 19%]
test/test_mps.py::TestConsistencyCPU::test_output_grad_match_nn_functional_pixel_unshuffle_cpu_float32 PASSED [0.0041s]                                                                                  [ 23%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_shuffle_cpu_bool PASSED [0.0355s]                                                                                            [ 28%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_shuffle_cpu_float16 PASSED [0.0025s]                                                                                         [ 33%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_shuffle_cpu_float32 PASSED [0.0027s]                                                                                         [ 38%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_shuffle_cpu_int16 PASSED [0.0375s]                                                                                           [ 42%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_shuffle_cpu_int32 PASSED [0.0357s]                                                                                           [ 47%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_shuffle_cpu_int64 PASSED [0.0360s]                                                                                           [ 52%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_shuffle_cpu_int8 PASSED [0.0346s]                                                                                            [ 57%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_shuffle_cpu_uint8 PASSED [0.0369s]                                                                                           [ 61%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_unshuffle_cpu_bool PASSED [0.0360s]                                                                                          [ 66%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_unshuffle_cpu_float16 PASSED [0.0023s]                                                                                       [ 71%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_unshuffle_cpu_float32 PASSED [0.0020s]                                                                                       [ 76%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_unshuffle_cpu_int16 PASSED [0.0371s]                                                                                         [ 80%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_unshuffle_cpu_int32 PASSED [0.0383s]                                                                                         [ 85%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_unshuffle_cpu_int64 PASSED [0.0357s]                                                                                         [ 90%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_unshuffle_cpu_int8 PASSED [0.0344s]                                                                                          [ 95%]
test/test_mps.py::TestConsistencyCPU::test_output_match_nn_functional_pixel_unshuffle_cpu_uint8 PASSED [0.0352s]                                                                                         [100%]

```

**Comment by alexdremov** [comment]
@kulinseth could you take a look please? Anything else to fix?
